routes/apis/contributors/share.js

```diff
@@ -0,0 +1,101 @@
+const { page_content_limit, ownDB, DB } = include('config/');
+const bcrypt = require('bcrypt');
+const crypto = require('crypto');
+const { email: sendEmail } = include('routes/helpers/');
+
+module.exports = async (req, res) => {
+  const { uuid, rights, username, email: c_email } = req.session || {};
+  let { pinboard, email } = Object.keys(req.query)?.length
+    ? req.query
+    : Object.keys(req.body)?.length
+    ? req.body
+    : {};
+
+  if (!uuid || !rights)
+    return res.status(401).json({ message: 'Unauthorized', success: false });
+
+  if (!pinboard)
+    return res.status(400).json({ message: 'Please provide a board ID', success: false });
+
+  if (!email)
+    return res.status(400).json({ message: 'Please provide an email.', success: false });
+
+  try {
+    const data = await DB.general.tx(async (t) => {
+      let user = await t.oneOrNone('SELECT * FROM users WHERE email = $1', [email]);
+
+      if (!user) {
+        if (!email.endsWith('@undp.org')) {
+          throw new Error('Invalid email. Only @undp.org emails are allowed.');
+        }
+
+        const generatePassword = () => crypto.randomBytes(8).toString('hex');
+
+        const autoGeneratedPassword = generatePassword();
+        const hashedPassword = bcrypt.hashSync(autoGeneratedPassword, 10);
+
+        const name = email.split('@')[0];
+
+        user = await t.one(
+          `INSERT INTO users (iso3, name, position, email, password, rights, confirmed, created_at) 
+           VALUES ($1, $2, $3, $4, $5, $6, $7, NOW()) 
+           RETURNING *`,
+          ['NUL', name, 'Contributor', email, hashedPassword, 1, false]
+        );
+      }
+
+      // Add user as a contributor
+      await t.none(
+        `INSERT INTO pinboard_contributors (participant, pinboard) 
+         VALUES ($1, $2) 
+         ON CONFLICT DO NOTHING`,
+        [user.uuid, pinboard]
+      );
+
+      // Fetch pinboard details
+      const board = await t.oneOrNone('SELECT title FROM pinboards WHERE id = $1', [pinboard]);
+
+      if (!board) {
+        throw new Error('Pinboard not found.');
+      }
+
+      // Send email notification
+      await sendEmail({
+        to: email,
+        cc: c_email,
+        subject: `You Have Been Added to the "${board.title}" board`,
```

Comment:
might be good to have a reminder here that this is on the sdg commons platform (although I guess it will be in the email address)

Reply:
also, maybe put everything in lowercase